File: Civitas ruby/lib/vista_textual.rb
#encoding:utf-8
require_relative "respuestas"
require_relative 'operaciones_juego'
require 'io/console'

module Civitas

  class Vista_textual

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    
    def comprar
      opciones = ["NO", "SI"]
      opcion = menu("¿Deseas comprar la casilla?", opciones)
      
      return $lista_Respuestas[opcion]
    end

    
    #Definicion de los atributos sin initilize?
    def gestionar
      opciones = ["VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA", "CONSTRUIR_CASA", "CONSTRUIR_HOTEL" , "TERMINAR", "REGALAR_CASA"]
      
      @iGestion = menu("¿Gestion inmobiliaria elegida?", opciones)
      @iPropiedad = @juegoModel.getJugadorActual.getNumCasillaActual
    end

    def getGestion
      return @iGestion
    end

    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      puts("\nSiguiente operacion: " + operacion.to_s)
    end

    def mostrarEventos
      while(Diario.instance.eventos_pendientes)
        evento = Diario.instance.leer_evento
        puts(evento)
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         @iGestion = -1
         @iPropiedad = -1
         actualizarVista
    end

    def actualizarVista
      @juegoModel.getJugadorActual.toString
      @juegoModel.getCasillaActual.toString
    end
    
    def salirCarcel
      opciones = ["Pagando", "Tirando el dado"]
      opcion = menu("Elige la forma para salir de la carcel", opciones)
      
      return $lista_Salidas[opcion]
    end
    
    def seleccionPropiedad
      propiedades = @juegoModel.getJugadorActual.getPropiedades
      opciones = []
      
      for nombre in propiedades
        opciones << nombre.getNombre
      end
      
      ip = menu("¿Que propiedad desea seleccionar?", opciones)
      
      return ip
    end

    
  end

end

File: Civitas java/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.GestionesInmobiliarias;
import civitas.Jugador;
import civitas.TituloPropiedad;

class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
      int opcion = menu ("¿Desea comprar la calle?",
              new ArrayList<> (Arrays.asList("NO", "SI")));
      return Respuestas.values()[opcion];
  }

  void gestionar () {
      ArrayList<String> lista = new ArrayList();
      lista.add("VENDER");
      lista.add("HIPOTECAR");
      lista.add("CANCELAR_HIPOTECA");
      lista.add("CONSTRUIR_CASA");
      lista.add("CONSTRUIR_HOTEL");
      lista.add("TERMINAR");
      lista.add("REGALAR CASA");
      
      iGestion = menu("¿Gestion inmobiliaria elegida?", lista);
      
      iPropiedad = juegoModel.getJugadorActual().getNumCasillaActual();
  }
  
  int seleccionPropiedad(){
      ArrayList<String> propiedades= new ArrayList();
      ArrayList<TituloPropiedad> titulos = juegoModel.getJugadorActual().getPropiedades();
      
      for(int i = 0; i < titulos.size(); i++)
        propiedades.add(titulos.get(i).getNombre());
      
      int ip = menu("¿Que propiedad desea seleccionar?", propiedades);
      
      return ip;
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.println("\nSiguiente operacion: " + operacion);
  }


  void mostrarEventos() {
      String evento;
      
      while(Diario.getInstance().eventosPendientes()){
          evento = Diario.getInstance().leerEvento();
          System.out.println(evento);
      }      
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        actualizarVista();
    }
  
  void actualizarVista(){
      juegoModel.getJugadorActual().toString();
      juegoModel.getCasillaActual().toString();
  } 
}
